src/wp-includes/option.php

```diff
@@ -1433,42 +1410,22 @@ function get_network_option( $network_id, $option, $default = false ) {
 		 * @param string $option     Option name.
 		 * @param int    $network_id ID of the network.
 		 */
-		return apply_filters( "default_site_option_{$option}", $default, $option, $network_id );
-	}
-
-	if ( ! is_multisite() ) {
-		/** This filter is documented in wp-includes/option.php */
-		$default = apply_filters( 'default_site_option_' . $option, $default, $option, $network_id );
+		$default = apply_filters( "default_site_option_{$option}", $default, $option, $network_id );
 		$value   = get_option( $option, $default );
 	} else {
-		$cache_key = "$network_id:$option";
-		$value     = wp_cache_get( $cache_key, 'site-options' );
-
-		if ( ! isset( $value ) || false === $value ) {
-			$row = $wpdb->get_row( $wpdb->prepare( "SELECT meta_value FROM $wpdb->sitemeta WHERE meta_key = %s AND site_id = %d", $option, $network_id ) );
-
-			// Has to be get_row() instead of get_var() because of funkiness with 0, false, null values.
-			if ( is_object( $row ) ) {
-				$value = $row->meta_value;
-				$value = maybe_unserialize( $value );
-				wp_cache_set( $cache_key, $value, 'site-options' );
-			} else {
-				if ( ! is_array( $notoptions ) ) {
-					$notoptions = array();
-				}
-
-				$notoptions[ $option ] = true;
-				wp_cache_set( $notoptions_key, $notoptions, 'site-options' );
-
-				/** This filter is documented in wp-includes/option.php */
-				$value = apply_filters( 'default_site_option_' . $option, $default, $option, $network_id );
+		$meta = get_metadata_raw( 'site', $network_id, $option );
+		if ( is_array( $meta ) && ! empty( $meta ) ) {
+			$value = array_shift( $meta );
+			if ( is_numeric( $value ) ) {
+				$value = +$value;
```

Comment:
@peterwilsoncc @felixarntz @desrosj This is a workaround to forces numbric values to be return an int / float. This should help with some edge casts and will mean lots less places where we need to cast to int.

Reply:
> his is a workaround to forces numbric values to be return an int / float.

Should these be forced to strings instead? For most requests numeric values are returns as strings.

```php
update_site_option( 'pwcc', 21 );
// Same request
var_dump( get_site_option( 'pwcc' ) ); // INT

// Next and subsequent requests
var_dump( get_site_option( 'pwcc' ) ); // String
```

Which is the bug, the first request or the subsequent requests?

Reply:
> subsequent requests

I am not seeing this. 
<img width="382" alt="Screenshot 2022-07-12 at 10 27 43" src="https://user-images.githubusercontent.com/237508/178458379-86742f4d-80eb-4c96-9bd3-52d5d9b866ff.png">

Reply:
> I am not seeing this.

I'm referring to the current behavior on trunk rather than this branch, ie the change on this branch causes a back-compat break.

![Screen Shot 2022-07-13 at 9 19 03 am](https://user-images.githubusercontent.com/519727/178614526-cd74ea8c-eb57-447c-a32f-74d441044946.png)

I think `get_site_option()` restoring a consistent type is a good thing; I guess the question is whether it is better to break back-compat on the initial or subsequent requests. My feeling is it is best to break the initial request as it's likely the string will be the most frequently returned type.

Reply:
@peterwilsoncc I am not seeing the same results as you. See this screenshot from trunk. 

<img width="464" alt="Screenshot 2022-07-19 at 11 05 39" src="https://user-images.githubusercontent.com/237508/179725615-76ed520c-49e8-4674-8f4f-d3f8294cc61a.png">


I think breaking backwards compat is a really big problem here. Consider the following code. 

```
if ( get_site_option( 'user_count' ) > 1000 ) {
```

or 

```
if ( get_site_option( 'user_count' ) === 1 ) {
```

Which would fail if typing is not correct. It seems like a very common thing to do, so store a number or a count in the site ( network ) option. We can't really break B/C here.

Reply:
@spacedmonkey The type change to string happens only on subsequent requests (or after exiting and restarting `wp shell`) 

When testing, you need to `exit` prior to `get_site_option()` to see the type change.

I resume this is why core has examples of casting in wrappers for getting options, eg in `get_user_count()`

https://github.com/WordPress/wordpress-develop/blob/2d9073413fa4714d13cc92075fcf4b9d7318777d/src/wp-includes/user.php#L1338

Reply:
@peterwilsoncc OHHHH. I finally get it. Okay, I understand now. Okay, I have added the casting back and tests are passsing. This will need a dev note, but I think we are good.